Bug in the ContainerResource implementation on the Containerd runtime.

### Discussed in https://github.com/open-telemetry/opentelemetry-java/discussions/4506

<div type='discussions-op-text'>

<sup>Originally posted by **obenkirane** May 30, 2022</sup>
Hello,
A Java exception is thrown at Spring Boot application launch; we do automatic instrumentation using the Javaagent.
We have found that this exception is thrown only when using **Containerd** runtime (no issue with Docker runtime).

Here is the thrown exception:

`[main] INFO io.opentelemetry.javaagent.tooling.VersionLogger - opentelemetry-javaagent - version: 1.14.0
OpenTelemetry Javaagent failed to start
java.lang.ExceptionInInitializerError
at io.opentelemetry.sdk.extension.resources.ContainerResourceProvider.createResource(ContainerResourceProvider.java:16)
at io.opentelemetry.sdk.autoconfigure.ResourceConfiguration.configureResource(ResourceConfiguration.java:50)
at io.opentelemetry.sdk.autoconfigure.AutoConfiguredOpenTelemetrySdkBuilder.build(AutoConfiguredOpenTelemetrySdkBuilder.java:308)
at io.opentelemetry.javaagent.tooling.OpenTelemetryInstaller.installOpenTelemetrySdk(OpenTelemetryInstaller.java:43)
at io.opentelemetry.javaagent.tooling.AgentInstaller.installBytebuddyAgent(AgentInstaller.java:115)
at io.opentelemetry.javaagent.tooling.AgentInstaller.installBytebuddyAgent(AgentInstaller.java:91)
at io.opentelemetry.javaagent.tooling.AgentStarterImpl.internalStart(AgentStarterImpl.java:98)
at io.opentelemetry.javaagent.tooling.AgentStarterImpl.start(AgentStarterImpl.java:71)
at io.opentelemetry.javaagent.bootstrap.AgentInitializer.initialize(AgentInitializer.java:35)
at io.opentelemetry.javaagent.OpenTelemetryAgent.startAgent(OpenTelemetryAgent.java:57)
at io.opentelemetry.javaagent.OpenTelemetryAgent.premain(OpenTelemetryAgent.java:45)
at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)
at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
at java.base/java.lang.reflect.Method.invoke(Unknown Source)
at java.instrument/sun.instrument.InstrumentationImpl.loadClassAndStartAgent(Unknown Source)
at java.instrument/sun.instrument.InstrumentationImpl.loadClassAndCallPremain(Unknown Source)
Caused by: java.lang.StringIndexOutOfBoundsException: begin 70, end 59, length 145
at java.base/java.lang.String.checkBoundsBeginEnd(Unknown Source)
at java.base/java.lang.String.substring(Unknown Source)
at io.opentelemetry.sdk.extension.resources.ContainerResource.getIdFromLine(ContainerResource.java:95)
at java.base/java.util.stream.ReferencePipeline$3$1.accept(Unknown Source)
at java.base/java.util.stream.ReferencePipeline$2$1.accept(Unknown Source)
at java.base/java.util.Spliterators$IteratorSpliterator.tryAdvance(Unknown Source)
at java.base/java.util.stream.ReferencePipeline.forEachWithCancel(Unknown Source)
at java.base/java.util.stream.AbstractPipeline.copyIntoWithCancel(Unknown Source)
at java.base/java.util.stream.AbstractPipeline.copyInto(Unknown Source)
at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(Unknown Source)
at java.base/java.util.stream.FindOps$FindOp.evaluateSequential(Unknown Source)
at java.base/java.util.stream.AbstractPipeline.evaluate(Unknown Source)
at java.base/java.util.stream.ReferencePipeline.findFirst(Unknown Source)
at io.opentelemetry.sdk.extension.resources.ContainerResource.extractContainerId(ContainerResource.java:72)
at io.opentelemetry.sdk.extension.resources.ContainerResource.buildResource(ContainerResource.java:39)
at io.opentelemetry.sdk.extension.resources.ContainerResource.buildSingleton(ContainerResource.java:34)
at io.opentelemetry.sdk.extension.resources.ContainerResource.<clinit>(ContainerResource.java:29)`

And here is an extract from the Dockerfile:

```
FROM adoptopenjdk/openjdk11:jre-11.0.11_9-alpine
RUN mkdir -p /app/bin
COPY target/opentelemetry-app-a.jar /app/bin
COPY target/opentelemetry-javaagent.jar /app/bin

ENV OTEL_OTLP_ENDPOINT http://opentelemetry-collector:4317
ENV OTEL_JAVAAGENT_DEBUG false

CMD java -Dotel.traces.exporter=otlp \
         -Dotel.metrics.exporter=none \
         -Dotel.logs.exporter=none \
         -javaagent:/app/bin/opentelemetry-javaagent.jar \
         -Dotel.exporter.otlp.endpoint=${OTEL_OTLP_ENDPOINT} \
         -Dotel.exporter.otlp.certificate=${OTEL_EXPORTER_OTLP_CERTIFICATE} \
         -Dotel.resource.attributes=service.name=${OTEL_SERVICE_NAME} \
         -Dapplication.home=/app/bin/ \
         -Dapplication.name=${OTEL_SERVICE_NAME} \
	 -Dotel.javaagent.debug=${OTEL_JAVAAGENT_DEBUG} \
	  -jar /app/bin/opentelemetry-app-a.jar
```
Thanks.</div>

--- a/sdk-extensions/resources/src/main/java/io/opentelemetry/sdk/extension/resources/ContainerResource.java
+++ b/sdk-extensions/resources/src/main/java/io/opentelemetry/sdk/extension/resources/ContainerResource.java
@@ -90,13 +90,17 @@ public final class ContainerResource {
     int startIdx = lastSection.lastIndexOf('-');
     int endIdx = lastSection.lastIndexOf('.');
 
-    if (startIdx > endIdx && endIdx != -1) {
+    if (startIdx == -1) {
+      startIdx = 0;
+    }
+    if (endIdx == -1) {
+      endIdx = lastSection.length();
+    }
+    if (startIdx > endIdx) {
       return null;
     }
 
-    String containerId =
-        lastSection.substring(
-            startIdx == -1 ? 0 : startIdx + 1, endIdx == -1 ? lastSection.length() : endIdx);
+    String containerId = lastSection.substring(startIdx, endIdx);
     if (OtelEncodingUtils.isValidBase16String(containerId) && !containerId.isEmpty()) {
       return containerId;
     } else {

--- a/sdk-extensions/resources/src/main/java/io/opentelemetry/sdk/extension/resources/ContainerResource.java
+++ b/sdk-extensions/resources/src/main/java/io/opentelemetry/sdk/extension/resources/ContainerResource.java
@@ -90,9 +90,7 @@ public final class ContainerResource {
     int startIdx = lastSection.lastIndexOf('-');
     int endIdx = lastSection.lastIndexOf('.');
 
-    if (startIdx == -1) {
-      startIdx = 0;
-    }
+    startIdx = startIdx == -1 ? 0 : startIdx + 1;
     if (endIdx == -1) {
       endIdx = lastSection.length();
     }

--- a/sdk-extensions/resources/src/main/java/io/opentelemetry/sdk/extension/resources/ContainerResource.java
+++ b/sdk-extensions/resources/src/main/java/io/opentelemetry/sdk/extension/resources/ContainerResource.java
@@ -9,7 +9,6 @@ import io.opentelemetry.api.common.Attributes;
 import io.opentelemetry.api.internal.OtelEncodingUtils;
 import io.opentelemetry.sdk.resources.Resource;
 import io.opentelemetry.semconv.resource.attributes.ResourceAttributes;
-import java.io.IOException;
 import java.nio.file.Files;
 import java.nio.file.Path;
 import java.nio.file.Paths;
@@ -73,8 +72,8 @@ public final class ContainerResource {
       if (value.isPresent()) {
         return value.get();
       }
-    } catch (IOException e) {
-      logger.log(Level.WARNING, "Unable to read file: " + e.getMessage());
+    } catch (Exception e) {
+      logger.log(Level.WARNING, "Unable to read file", e);
     }
     return null;
   }
@@ -90,6 +89,10 @@ public final class ContainerResource {
     String lastSection = line.substring(lastSlashIdx + 1);
     int startIdx = lastSection.lastIndexOf('-');
     int endIdx = lastSection.lastIndexOf('.');
+
+    if (startIdx > endIdx && endIdx != -1) {
+      return null;
+    }
 
     String containerId =
         lastSection.substring(

--- a/sdk-extensions/resources/src/test/java/io/opentelemetry/sdk/extension/resources/ContainerResourceTest.java
+++ b/sdk-extensions/resources/src/test/java/io/opentelemetry/sdk/extension/resources/ContainerResourceTest.java
@@ -17,47 +17,22 @@ import java.nio.file.Path;
 import org.junit.jupiter.api.Test;
 import org.junit.jupiter.api.io.TempDir;
 
-public class ContainerResourceTest {
-
-  // Invalid because ID is not a hex string
-  private static final String INVALID_CGROUP_LINE_1 =
-      "13:name=systemd:/podruntime/docker/kubepods/ac679f8a8319c8cf7d38e1adf263bc08d23zzzz";
-
-  // with suffix
-  private static final String CGROUP_LINE_1 =
-      "13:name=systemd:/podruntime/docker/kubepods/ac679f8a8319c8cf7d38e1adf263bc08d23.aaaa";
-  private static final String EXPECTED_CGROUP_1 = "ac679f8a8319c8cf7d38e1adf263bc08d23";
-
-  // with prefix and suffix
-  private static final String CGROUP_LINE_2 =
-      "13:name=systemd:/podruntime/docker/kubepods/crio-dc679f8a8319c8cf7d38e1adf263bc08d23.stuff";
-  private static final String EXPECTED_CGROUP_2 = "dc679f8a8319c8cf7d38e1adf263bc08d23";
-
-  // just container id
-  private static final String CGROUP_LINE_3 =
-      "13:name=systemd:/pod/d86d75589bf6cc254f3e2cc29debdf85dde404998aa128997a819ff991827356";
-  private static final String EXPECTED_CGROUP_3 =
-      "d86d75589bf6cc254f3e2cc29debdf85dde404998aa128997a819ff991827356";
-
-  // with prefix
-  private static final String CGROUP_LINE_4 =
-      "//\n"
-          + "1:name=systemd:/podruntime/docker/kubepods/docker-dc579f8a8319c8cf7d38e1adf263bc08d23"
-          + "2:name=systemd:/podruntime/docker/kubepods/docker-dc579f8a8319c8cf7d38e1adf263bc08d23"
-          + "3:name=systemd:/podruntime/docker/kubepods/docker-dc579f8a8319c8cf7d38e1adf263bc08d23";
-
-  private static final String EXPECTED_CGROUP_4 = "dc579f8a8319c8cf7d38e1adf263bc08d23";
-
-  // with two dashes in prefix
-  private static final String CGROUP_LINE_5 =
-      "11:perf_event:/kubepods.slice/kubepods-burstable.slice/kubepods-burstable-pod4415fd05_2c0f_4533_909b_f2180dca8d7c.slice/cri-containerd-713a77a26fe2a38ebebd5709604a048c3d380db1eb16aa43aca0b2499e54733c.scope";
-  private static final String EXPECTED_CGROUP_5 =
-      "713a77a26fe2a38ebebd5709604a048c3d380db1eb16aa43aca0b2499e54733c";
+class ContainerResourceTest {
 
   @Test
-  public void testNegativeCases(@TempDir Path tempFolder) throws IOException {
+  void buildResource_Invalid(@TempDir Path tempFolder) throws IOException {
     // invalid containerId (non-hex)
-    Path cgroup = createCGroup(tempFolder.resolve("cgroup1"), INVALID_CGROUP_LINE_1);
+    Path cgroup =
+        createCGroup(
+            tempFolder.resolve("cgroup1"),
+            "13:name=systemd:/podruntime/docker/kubepods/ac679f8a8319c8cf7d38e1adf263bc08d23zzzz");
+    assertThat(buildResource(cgroup)).isEqualTo(Resource.empty());
+
+    // unrecognized format (last "-" is after last ".")
+    cgroup =
+        createCGroup(
+            tempFolder.resolve("cgroup1"),
+            "13:name=systemd:/podruntime/docker/kubepods/ac679f8.a8319c8cf7d38e1adf263bc08-d23zzzz");
     assertThat(buildResource(cgroup)).isEqualTo(Resource.empty());
 
     // test invalid file
@@ -66,21 +41,49 @@ public class ContainerResourceTest {
   }
 
   @Test
-  public void testContainer(@TempDir Path tempFolder) throws IOException {
-    Path cgroup = createCGroup(tempFolder.resolve("cgroup1"), CGROUP_LINE_1);
-    assertThat(getContainerId(buildResource(cgroup))).isEqualTo(EXPECTED_CGROUP_1);
+  void buildResource_Valid(@TempDir Path tempFolder) throws IOException {
+    // with suffix
+    Path cgroup =
+        createCGroup(
+            tempFolder.resolve("cgroup1"),
+            "13:name=systemd:/podruntime/docker/kubepods/ac679f8a8319c8cf7d38e1adf263bc08d23.aaaa");
+    assertThat(getContainerId(buildResource(cgroup)))
+        .isEqualTo("ac679f8a8319c8cf7d38e1adf263bc08d23");
 
-    Path cgroup2 = createCGroup(tempFolder.resolve("cgroup2"), CGROUP_LINE_2);
-    assertThat(getContainerId(buildResource(cgroup2))).isEqualTo(EXPECTED_CGROUP_2);
+    // with prefix and suffix
+    Path cgroup2 =
+        createCGroup(
+            tempFolder.resolve("cgroup2"),
+            "13:name=systemd:/podruntime/docker/kubepods/crio-dc679f8a8319c8cf7d38e1adf263bc08d23.stuff");
+    assertThat(getContainerId(buildResource(cgroup2)))
+        .isEqualTo("dc679f8a8319c8cf7d38e1adf263bc08d23");
 
-    Path cgroup3 = createCGroup(tempFolder.resolve("cgroup3"), CGROUP_LINE_3);
-    assertThat(getContainerId(buildResource(cgroup3))).isEqualTo(EXPECTED_CGROUP_3);
+    // just container id
+    Path cgroup3 =
+        createCGroup(
+            tempFolder.resolve("cgroup3"),
+            "13:name=systemd:/pod/d86d75589bf6cc254f3e2cc29debdf85dde404998aa128997a819ff991827356");
+    assertThat(getContainerId(buildResource(cgroup3)))
+        .isEqualTo("d86d75589bf6cc254f3e2cc29debdf85dde404998aa128997a819ff991827356");
 
-    Path cgroup4 = createCGroup(tempFolder.resolve("cgroup4"), CGROUP_LINE_4);
-    assertThat(getContainerId(buildResource(cgroup4))).isEqualTo(EXPECTED_CGROUP_4);
+    // with prefix
+    Path cgroup4 =
+        createCGroup(
+            tempFolder.resolve("cgroup4"),
+            "//\n"
+                + "1:name=systemd:/podruntime/docker/kubepods/docker-dc579f8a8319c8cf7d38e1adf263bc08d23"
+                + "2:name=systemd:/podruntime/docker/kubepods/docker-dc579f8a8319c8cf7d38e1adf263bc08d23"
+                + "3:name=systemd:/podruntime/docker/kubepods/docker-dc579f8a8319c8cf7d38e1adf263bc08d23");
+    assertThat(getContainerId(buildResource(cgroup4)))
+        .isEqualTo("dc579f8a8319c8cf7d38e1adf263bc08d23");
 
-    Path cgroup5 = createCGroup(tempFolder.resolve("cgroup5"), CGROUP_LINE_5);
-    assertThat(getContainerId(buildResource(cgroup5))).isEqualTo(EXPECTED_CGROUP_5);
+    // with two dashes in prefix
+    Path cgroup5 =
+        createCGroup(
+            tempFolder.resolve("cgroup5"),
+            "11:perf_event:/kubepods.slice/kubepods-burstable.slice/kubepods-burstable-pod4415fd05_2c0f_4533_909b_f2180dca8d7c.slice/cri-containerd-713a77a26fe2a38ebebd5709604a048c3d380db1eb16aa43aca0b2499e54733c.scope");
+    assertThat(getContainerId(buildResource(cgroup5)))
+        .isEqualTo("713a77a26fe2a38ebebd5709604a048c3d380db1eb16aa43aca0b2499e54733c");
   }
 
   private static String getContainerId(Resource resource) {

